cpp/src/io/parquet/page_data.cu

```diff
@@ -1106,69 +1600,208 @@ extern "C" __global__ void __launch_bounds__(NTHREADS) gpuDecodePageData(
-      if (out_pos < target_pos && row_idx >= 0 && s->first_row + row_idx < s->num_rows) {
+      uint32_t src_pos = out_pos + skipped_leaf_values;
+
+      int output_value_idx = s->nz_idx[rolling_index(out_pos)];
+
+      if (out_pos < target_pos && output_value_idx >= 0 && output_value_idx < s->num_input_values) {
+        // nesting level that is storing actual leaf values
+        int leaf_level_index = s->col.max_level[level_type::REPETITION];
+
         uint32_t dtype_len = s->dtype_len;
-        uint8_t *dst       = s->data_out + (size_t)row_idx * dtype_len;
+        uint8_t *dst       = s->page.nesting[leaf_level_index].data_out +
+                       static_cast<size_t>(output_value_idx) * dtype_len;
         if (dtype == BYTE_ARRAY)
-          gpuOutputString(s, out_pos, dst);
+          gpuOutputString(s, src_pos, dst);
         else if (dtype == BOOLEAN)
-          gpuOutputBoolean(s, out_pos, dst);
+          gpuOutputBoolean(s, src_pos, dst);
         else if (s->col.converted_type == DECIMAL)
-          gpuOutputDecimal(s, out_pos, reinterpret_cast<double *>(dst), dtype);
+          gpuOutputDecimal(s, src_pos, reinterpret_cast<double *>(dst), dtype);
         else if (dtype == INT96)
-          gpuOutputInt96Timestamp(s, out_pos, reinterpret_cast<int64_t *>(dst));
+          gpuOutputInt96Timestamp(s, src_pos, reinterpret_cast<int64_t *>(dst));
         else if (dtype_len == 8) {
           if (s->ts_scale)
-            gpuOutputInt64Timestamp(s, out_pos, reinterpret_cast<int64_t *>(dst));
+            gpuOutputInt64Timestamp(s, src_pos, reinterpret_cast<int64_t *>(dst));
           else
-            gpuOutputFast(s, out_pos, reinterpret_cast<uint2 *>(dst));
+            gpuOutputFast(s, src_pos, reinterpret_cast<uint2 *>(dst));
         } else if (dtype_len == 4)
-          gpuOutputFast(s, out_pos, reinterpret_cast<uint32_t *>(dst));
+          gpuOutputFast(s, src_pos, reinterpret_cast<uint32_t *>(dst));
         else
-          gpuOutputGeneric(s, out_pos, dst, dtype_len);
+          gpuOutputGeneric(s, src_pos, dst, dtype_len);
       }
+
       if (t == out_thread0) { *(volatile int32_t *)&s->out_pos = target_pos; }
     }
     __syncthreads();
   }
-  __syncthreads();
-  if (!t) {
-    // Update the number of rows (after cropping to [min_row, min_row+num_rows-1]), and number of
-    // valid values
-    pages[page_idx].num_rows    = s->num_rows - s->first_row;
-    pages[page_idx].valid_count = (s->error) ? -s->error : s->page.valid_count;
+}
+
+struct chunk_row_output_iter {
+  PageInfo *p;
+  using value_type        = size_type;
+  using difference_type   = size_type;
+  using pointer           = size_type *;
+  using reference         = size_type &;
+  using iterator_category = thrust::output_device_iterator_tag;
+
+  chunk_row_output_iter operator+ __host__ __device__(int i)
+  {
+    return chunk_row_output_iter{p + i};
+  }
+  chunk_row_output_iter operator[] __host__ __device__(int i)
```

Comment:
shouldn't this return a `value_type`?

Reply:
So I am completely baffled at what's going on here.  I think the right thing is actually a reference, not a value_type. Ie

```
reference operator[] __host__ __device__(int i)
{
  return p[i].chunk_row;
}
```
This also compiles, runs, and produces the right output, even though it is returning a completely different thing that the existing operator[] (which also compiles, runs, and produces the right output).   And I've verified this actually gets hit at runtime. 

Pinging @jrhemstad  and @codereport for some deeper iterator insights here.

Reply:
Right, it should be a reference, my bad.
I assumed it's not really used, real strange that it works either way.

Reply:
I remember looking up how to model an OutputIterator since I haven't done it before.  Then I got the unexpected compiler error because it was missing that operator.  Added it (clearly did it wrong) - compiled, got what I was expecting for results.  What really has me confused is that that operator actually gets called, and it both compiles and works with either of these definitions, which are completely different!   I'll change it to use the reference but man would I like to know what's going on.

Reply:
Fwiw, compiler error is:
```
cpp/build/_deps/cub-src/cub/block/block_store.cuh(295): error: no operator "[]" matches these operands
            operand types are: cudf::io::parquet::gpu::chunk_row_output_iter [ int ]
          detected during:
            instantiation of "void cub::StoreDirectWarpStriped(int, OutputIteratorT, T (&)[ITEMS_PER_THREAD]) [with T=cudf::io::parquet::gpu::chunk_row_output_iter::value_type, ITEMS_PER_THREAD=9, OutputIteratorT=cudf::io::parquet::gpu::chunk_row_output_iter]" 
(737): here

```
So clearly it's being used.

Reply:
Fixed this.  I also mocked it up using the thrust::iterator_facade, but I'd prefer to keep it as is to avoid changing things too drastically at the last second.